crates/corro-agent/src/agent/util.rs

```diff
@@ -0,0 +1,1479 @@
+}
+
+// DOCME: provide some context for this function
+// TODO: move to a more appropriate module?
+pub async fn sync_loop(
+    agent: Agent,
+    transport: Transport,
+    mut rx_apply: Receiver<(ActorId, Version)>,
+    mut tripwire: Tripwire,
+) {
+    let mut sync_backoff = backoff::Backoff::new(0)
+        .timeout_range(Duration::from_secs(1), MAX_SYNC_BACKOFF)
+        .iter();
+    let next_sync_at = tokio::time::sleep(sync_backoff.next().unwrap());
+    tokio::pin!(next_sync_at);
+
+    loop {
+        enum Branch {
+            Tick,
+            BackgroundApply { actor_id: ActorId, version: Version },
+        }
+
+        let branch = tokio::select! {
+            biased;
+
+            maybe_item = rx_apply.recv() => match maybe_item {
+                Some((actor_id, version)) => Branch::BackgroundApply{actor_id, version},
+                None => {
+                    debug!("background applies queue is closed, breaking out of loop");
+                    break;
+                }
+            },
+
+            _ = &mut next_sync_at => {
+                Branch::Tick
+            },
+            _ = &mut tripwire => {
+                break;
+            }
+        };
+
+        match branch {
+            Branch::Tick => {
+                // ignoring here, there is trying and logging going on inside
+                match handlers::handle_sync(&agent, &transport)
+                    .preemptible(&mut tripwire)
+                    .await
+                {
+                    tripwire::Outcome::Preempted(_) => {
+                        warn!("aborted sync by tripwire");
+                        break;
+                    }
+                    tripwire::Outcome::Completed(res) => {
+                        if let Err(e) = res {
+                            error!("could not sync: {e}");
+                            // keep syncing until we successfully sync
+                            continue;
+                        }
+                    }
+                }
+                next_sync_at
+                    .as_mut()
+                    .reset(tokio::time::Instant::now() + sync_backoff.next().unwrap());
+            }
+            Branch::BackgroundApply { actor_id, version } => {
+                debug!(%actor_id, %version, "picked up background apply of buffered changes");
+                match process_fully_buffered_changes(&agent, actor_id, version).await {
+                    Ok(false) => {
+                        warn!(%actor_id, %version, "did not apply buffered changes");
+                    }
+                    Ok(true) => {
+                        debug!(%actor_id, %version, "succesfully applied buffered changes");
+                    }
+                    Err(e) => {
+                        error!(%actor_id, %version, "could not apply fully buffered changes: {e}");
+                    }
+                }
+            }
+        }
+    }
+}
+
+// DOCME: provide some context for this function
```

Comment:
Compacts the database by finding cleared versions. One actor at a time to lock for a minimal amount of time.